src/Lib.P2P/Swarm.cs

```diff
@@ -0,0 +1,989 @@
+                LocalPeer = LocalPeer,
+
+                // TODO: LocalAddress
+                LocalPeerKey = LocalPeerKey,
+                RemotePeer = remote,
+                RemoteAddress = addr,
+                Stream = stream
+            };
+
+            // Are we communicating to a private network?
+            if (NetworkProtector != null)
+                connection.Stream = await NetworkProtector.ProtectAsync(connection).ConfigureAwait(false);
+
+            return connection;
+        }
+
+        /// <summary>
+        ///   Disconnect from a peer.
+        /// </summary>
+        /// <param name="address">
+        ///   An ipfs <see cref="MultiAddress"/>, such as
+        ///  <c>/ip4/104.131.131.82/tcp/4001/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ</c>.
+        /// </param>
+        /// <param name="cancel">
+        ///   Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException"/> is raised.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation.
+        /// </returns>
+        /// <remarks>
+        ///   If the peer is not conected, then nothing happens.
+        /// </remarks>
+        public Task DisconnectAsync(MultiAddress address, CancellationToken cancel = default(CancellationToken))
+        {
+            Manager.Remove(address.PeerId);
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        ///   Start listening on the specified <see cref="MultiAddress"/>.
+        /// </summary>
+        /// <param name="address">
+        ///   Typically "/ip4/0.0.0.0/tcp/4001" or "/ip6/::/tcp/4001".
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation.  The task's result
+        ///   is a <see cref="MultiAddress"/> than can be used by another peer
+        ///   to connect to tis peer.
+        /// </returns>
+        /// <exception cref="Exception">
+        ///   Already listening on <paramref name="address"/>.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        ///   <paramref name="address"/> is missing a transport protocol (such as tcp or udp).
+        /// </exception>
+        /// <remarks>
+        ///   Allows other peers to <see cref="ConnectAsync(MultiAddress, CancellationToken)">connect</see>
+        ///   to the <paramref name="address"/>.
+        ///   <para>
+        ///   The <see cref="Peer.Addresses"/> of the <see cref="LocalPeer"/> are updated.  If the <paramref name="address"/> refers to
+        ///   any IP address ("/ip4/0.0.0.0" or "/ip6/::") then all network interfaces addresses
+        ///   are added.  If the port is zero (as in "/ip6/::/tcp/0"), then the peer addresses contains the actual port number
+        ///   that was assigned.
+        ///   </para>
+        /// </remarks>
+        public Task<MultiAddress> StartListeningAsync(MultiAddress address)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The Cyclomatic Complexity of this method is 12 which is greater than 10 authorized.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)